queueing/inserter.py

```diff
@@ -25,40 +31,58 @@ def __init__(self, client, database_name, collection_name, unique):
 
         logging.info("New Inserter created for {}.".format(self.identifier))
 
-    def __insert_items(self):
-        response = self.__client.insert_many_to_collection(
-            self.database_name, self.collection_name, self.__items
+    def __handle_insertion_error(self, response, items):
+        logging.warning(
+            "The exception {} occured during the insertion of {} items in {}.".format(
+                response.message, len(items), self.identifier
+            )
         )
-        logging.debug(
-            "{} documents inserted in {}.".format(len(self.__items), self.identifier)
+        for item in items:
+            if response.need_upsert:
+                item["need_upsert"] = "True"
+            kafka_producer.send(self.topic, value=item)
+
+    def __insert_items(self, reason=""):
+        items = [copy(item["payload"]) for item in self.__items]
+        response = self.__client.insert_many_to_collection(
+            self.database_name, self.collection_name, items
         )
+        if response.ok:
+            logging.info(
+                "{} documents inserted [{}] in {}.".format(
+                    len(self.__items), reason, self.identifier
+                )
+            )
+        else:
+            self.__handle_insertion_error(response, self.__items)
         self.__items = []
-        return response
 
     def is_inactive(self):
         return time.time() - self.__last_activity > ACTIVITY_TIME_THRESHOLD
 
     def insert(self, item):
-        if self.unique:
-            self.__client.insert_one_to_unique_collection(
-                self.database_name, self.collection_name, item
+        if self.unique or item.get("need_upsert") is not None:
+            response = self.__client.insert_one_to_unique_collection(
+                self.database_name, self.collection_name, copy(item["payload"])
             )
-            logging.debug("1 document inserted in {}.".format(self.identifier))
+            if response.ok:
+                logging.debug("1 document inserted in {}.".format(self.identifier))
+            else:
+                self.__handle_insertion_error(response, [item])
         else:
             with self.__lock:
                 self.__items.append(item)
-                if (
-                    sys.getsizeof(self.__items) > BATCH_SIZE_THRESHOLD
-                    or time.time() - self.__last_insertion > INSERT_TIME_THRESHOLD
-                ):
-                    response = self.__insert_items()
+                if sys.getsizeof(self.__items) > BATCH_SIZE_THRESHOLD:
+                    self.__insert_items("size threshold")
+                elif time.time() - self.__last_insertion > INSERT_TIME_THRESHOLD:
+                    self.__insert_items("time threshold")
 
         self.__last_insertion = time.time()
         self.__last_activity = time.time()
 
-    def flush(self):
+    def flush(self, reason="unknown"):
```

Comment:
When would the `"unknown"` reason be used?

Reply:
It should be a required field, changed